### PR TITLE
[Feature] Improve the observability of integration tests

### LINKED ISSUE #766: [Feature] Improve the observability of integration tests
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

We can reuse the function [show_cluster_info()](https://github.com/ray-project/kuberay/blob/49c44bff5a807a680248b30b5563f985a075be9b/tests/framework/prototype.py#L120) to get more information when CI tests fail. To achieve this goal, we need to:

(1) `show_cluster_info()` needs to support operator installed by Helm. The operator installed by Helm does not have the label `app.kubernetes.io/component`. It has `app.kubernetes.io/instance` instead.

(2) Check if the test passes or not in [tearDown()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown). Call `show_cluster_info()` if the test fails.

### Use case

Observability improvement can help us debug CI failures. 

### Related issues

#759 

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jasoonn**
## Why are these changes needed?
Add ```app.kubernetes.io/component: kuberay-operator``` label in [helm-chart/kuberay-operator/templates/deployment.yaml](https://github.com/ray-project/kuberay/blob/5177ebfbc885faa4abca8109f1fb692b3c85e880/helm-chart/kuberay-operator/templates/deployment.yaml) to make operator pod can be selected in ```show_cluster_info()```.

### Why not call ```show_cluster_info()``` in [tearDown()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown)?
The only way I found to check if the test pass in ```teardown()``` is to access the private field ```self._outcome``` in unittest.TestCase object([reference](https://stackoverflow.com/questions/4414234/getting-pythons-unittest-results-in-a-teardown-method)). The fields in ```self._outcome``` changes in Python 3.11. It seems like accessing ```self._outcome``` in unittest.TestCase is not a robust approach.

### Why added  hard code label ```app.kubernetes.io/component=kuberay-operator```?
Because all of the labels app.kubernetes.io/component: for the operator is hard code with value kuberay-operato, I added app.kubernetes.io/component: kuberay-operator to the helm chart to make ```show_cluster_info``` able to get logs from the operator pod.

## Related issue number
Closes #766 

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: tests/framework/prototype.py [hunk 1]
```diff
@@ -201,9 +207,9 @@ def show_cluster_info(cr_namespace):
     # -1 when using selector.
     shell_subprocess_run(f'kubectl logs -n={cr_namespace} -l ray.io/node-type=head --tail=-1')
     operator_namespace = shell_subprocess_check_output('kubectl get pods '
-        '-l app.kubernetes.io/component=kuberay-operator -A '
+        '-l app.kubernetes.io/instance=kuberay-operator -A '
```
Comment (kevin85421) [inline tests/framework/prototype.py:210]: We need to make this function compatible with both operator deployment methods (kubectl & helm).

File: tests/framework/prototype.py [hunk 2]
```diff
@@ -146,13 +146,19 @@ def check_pod_running(pods) -> bool:
             return False
     return True
 
-def shell_subprocess_run(command, check = True):
+def shell_subprocess_run(command, cr_namespace = None, check = True):
```
Comment (kevin85421) [inline tests/framework/prototype.py:149]: This framework is possible to use for other operators, and thus maybe we should not couple them together.

**Review by kevin85421 — COMMENTED** [review]
Thank you for this contribution!

File: tests/framework/prototype.py (hunk 2 above)
Comment (DmitriGekhtman) [reply tests/framework/prototype.py:149]: I agree -- K8s concepts shouldn't appear in a command utility.

File: tests/framework/prototype.py (hunk 1 above)
Comment (DmitriGekhtman) [reply tests/framework/prototype.py:210]: Feel free to add a label to the helm chart.

Comment (jasoonn) [reply tests/framework/prototype.py]: Thanks for the suggestion. I have added a label to the helm chart.

**Comment by jasoonn** [comment]
@kevin85421 @DmitriGekhtman Thanks for the suggestion. I have updated my pr.

**Comment by DmitriGekhtman** [comment]
Was there a Git issue? I think we lost most of the changes (I see `Files changed 1`)

File: tests/framework/prototype.py (hunk 1 above)
Comment (jasoonn) [reply tests/framework/prototype.py]: I found all of the labels ```app.kubernetes.io/component:``` for the operator is hard code with value ```kuberay-operator```.  Hence, I added ```app.kubernetes.io/component: kuberay-operator``` to the helm chart.

**Comment by jasoonn** [comment]
@DmitriGekhtman Because the label utilized in the original ```show_cluster_info()``` is ```app.kubernetes.io/component: kuberay-operator```, we only need to add that label in the helm chart. Besides, I changed the ```shell_subprocess_run()``` back.

**Comment by jasoonn** [comment]
To improve the observability of integration tests, I called ```show_cluster_info("default")``` in several places in [tests/compatibility-test.py](https://github.com/ray-project/kuberay/blob/master/tests/compatibility-test.py). The places I choose are where the most CI tests failed in: execution of script files. e.g. [test_ray_serve](https://github.com/ray-project/kuberay/blob/f0abc1d1fa7c2ce1bff3781a6127ce74033ea04a/tests/compatibility-test.py#L198-L202) and [test_detach_actor](https://github.com/ray-project/kuberay/blob/f0abc1d1fa7c2ce1bff3781a6127ce74033ea04a/tests/compatibility-test.py#L239-L244).

**Review by DmitriGekhtman — APPROVED** [review]

**Comment by DmitriGekhtman** [comment]
LGTM, I'll merge this after the CI finishes running.
This will be useful to have in the release branch.

File: tests/compatibility-test.py [hunk 3]
```diff
@@ -198,6 +193,7 @@ def test_ray_serve(self):
         exit_code, _ = utils.exec_run_container(container, f'python3 /usr/local/test_ray_serve_1.py {ray_namespace}', timeout_sec = 180)
 
         if exit_code != 0:
+            show_cluster_info("default")
```
Comment (kevin85421) [inline tests/compatibility-test.py:196]: I found that we hard-coded a lot of "default" in both `test_ray_serve` and `test_detached_actor`. How about creating a local variable in each test to replace this hard-coded method?

For example,

`utils.get_pod(namespace='default', label_selector='ray.io/node-type=head')`

File: tests/compatibility-test.py [hunk 4]
```diff
@@ -177,13 +178,7 @@ def test_kill_head(self):
...
         if rtn != 0:
-            shell_subprocess_run('kubectl get pods -A')
-            shell_subprocess_run(
-                'kubectl describe pod $(kubectl get pods | grep -e "-head" | awk "{print \$1}")')
-            shell_subprocess_run(
-                'kubectl logs $(kubectl get pods | grep -e "-head" | awk "{print \$1}")')
-            shell_subprocess_run(
-                'kubectl logs -n $(kubectl get pods -A | grep -e "-operator" | awk \'{print $1 "  " $2}\')')
+            show_cluster_info("default")
         assert rtn == 0
```
Comment (kevin85421) [inline tests/compatibility-test.py:182]: Remove the assert and raise an exception in the `if` statement?

File: tests/framework/prototype.py (hunk 1 above)
Comment (kevin85421) [reply tests/framework/prototype.py]: > I found all of the labels `app.kubernetes.io/component:` for the operator is hard code with value `kuberay-operator`. Hence, I added `app.kubernetes.io/component: kuberay-operator` to the helm chart.

Thank @jasoonn for the explanation! Can you add this statement to your PR description? Then, you can resolve this conversation.

File: tests/compatibility-test.py (hunk 4 above)
Comment (jasoonn) [reply tests/compatibility-test.py]: Updated https://github.com/ray-project/kuberay/pull/775/commits/7d640b0e8cacefef658aed7ee1cabf6431a3eb3d.

File: tests/compatibility-test.py (hunk 3 above)
Comment (jasoonn) [reply tests/compatibility-test.py]: I have updated in https://github.com/ray-project/kuberay/pull/775/commits/d5c8fd6eb15614893f700495bf03bd7a4ff26a33 and https://github.com/ray-project/kuberay/pull/775/commits/c0d8ef65846495b62375e3783239e9ded2ea8722. To be noted, these two functions will utilize the same raycluster which is created in ```setUpClass()```. If we want to separate the namespace in the future, we will need to move the creation of raycluster into corresponding functions.

**Comment by jasoonn** [comment]
@kevin85421 Thanks for the suggestion. I have updated the pr.

File: tests/compatibility-test.py [hunk 5]
```diff
@@ -177,16 +178,11 @@ def test_kill_head(self):
...
-            shell_subprocess_run(
-                'kubectl logs $(kubectl get pods | grep -e "-head" | awk "{print \$1}")')
-            shell_subprocess_run(
-                'kubectl logs -n $(kubectl get pods -A | grep -e "-operator" | awk \'{print $1 "  " $2}\')')
-        assert rtn == 0
+            show_cluster_info("default")
+            raise Exception(f"Nonzero return code {rtn} in test_kill_head()")
 
     def test_ray_serve(self):
+        cluster_name_space = "default"
```
Comment (kevin85421) [inline tests/compatibility-test.py:185]:
```suggestion
        cluster_namespace = "default"
```

Comment (kevin85421) [inline tests/compatibility-test.py:185]: `cluster_namespace`

Comment (jasoonn) [reply tests/compatibility-test.py]: Update in https://github.com/ray-project/kuberay/pull/775/commits/bc948f26fa65439f730da70884b7d1328cd1b164.

**Comment by jasoonn** [comment]
@kevin85421 Thanks for the suggestion. I have updated the pr.

**Review by kevin85421 — APPROVED** [review]
LGTM. Wait for CI to pass.

**Review by DmitriGekhtman — APPROVED** [review]

**Pull request merged**